pkg/machine/compression/decompress.go

```diff
@@ -1,313 +1,113 @@
+	// Are we reading full image file?
+	// Only few bytes are read to detect
+	// the compression type
+	compressedFileContent, err := compressedVMFile.Read()
 	if err != nil {
 		return err
 	}
-	file, err := os.Open(src)
-	if err != nil {
-		return err
-	}
-	defer file.Close()
-
-	p, bar := utils.ProgressBar(prefix, stat.Size(), prefix+": done")
-	proxyReader := bar.ProxyReader(file)
-	defer func() {
-		if err := proxyReader.Close(); err != nil {
-			logrus.Error(err)
-		}
-	}()
 
-	// Prefer Xz utils for fastest performance, fallback to go xi2 impl
-	if _, err := exec.LookPath("xz"); err == nil {
-		cmd = exec.Command("xz", "-d", "-c")
-		cmd.Stdin = proxyReader
-		read, err = cmd.StdoutPipe()
-		if err != nil {
-			return err
-		}
-		cmd.Stderr = os.Stderr
-	} else {
-		// This XZ implementation is reliant on buffering. It is also 3x+ slower than XZ utils.
-		// Consider replacing with a faster implementation (e.g. xi2) if podman machine is
-		// updated with a larger image for the distribution base.
-		buf := bufio.NewReader(proxyReader)
-		read, err = xz.NewReader(buf)
-		if err != nil {
-			return err
-		}
+	var d decompressor
+	if d, err = newDecompressor(compressedFilePath, compressedFileContent); err != nil {
+		return err
 	}
 
-	done := make(chan bool)
-	go func() {
-		if _, err := io.Copy(output, read); err != nil {
-			logrus.Error(err)
-		}
-		output.Close()
-		done <- true
-	}()
-
-	if cmd != nil {
-		err := cmd.Start()
-		if err != nil {
-			return err
-		}
-		p.Wait()
-		return cmd.Wait()
-	}
-	<-done
-	p.Wait()
-	return nil
+	return runDecompression(d, decompressedFilePath)
 }
 
-func decompressEverythingElse(prefix string, src string, output io.WriteCloser) error {
-	stat, err := os.Stat(src)
-	if err != nil {
-		return err
-	}
-	f, err := os.Open(src)
-	if err != nil {
-		return err
-	}
-	p, bar := utils.ProgressBar(prefix, stat.Size(), prefix+": done")
-	proxyReader := bar.ProxyReader(f)
-	defer func() {
-		if err := proxyReader.Close(); err != nil {
-			logrus.Error(err)
-		}
-	}()
-	uncompressStream, _, err := compression.AutoDecompress(proxyReader)
-	if err != nil {
-		return err
+func newDecompressor(compressedFilePath string, compressedFileContent []byte) (decompressor, error) {
+	compressionType := archive.DetectCompression(compressedFileContent)
+	os := runtime.GOOS
+	hasZipSuffix := strings.HasSuffix(compressedFilePath, zipExt)
+
+	switch {
+	case compressionType == archive.Xz:
+		return newXzDecompressor(compressedFilePath)
+	// Zip files are not guaranteed to have a magic number at the beginning
+	// of the file, so we need to use the file name to detect them.
+	case compressionType == archive.Uncompressed && hasZipSuffix:
+		return newZipDecompressor(compressedFilePath)
+	case compressionType == archive.Uncompressed:
+		return newUncompressedDecompressor(compressedFilePath)
+	// Using special compressors on MacOS because default ones
+	// in c/image/pkg/compression are slow with sparse files.
```

Comment:
AFAICT the implementation is now the same, the difference is _only_ that these two use `sparseWriter`. Am I missing something?

The previous version said “macOS gzipped VM images are sparse”.

And https://github.com/containers/podman/pull/21592#discussion_r1495889105 suggested that previously we had to choose between sparse processing and fast decompression, so we choose the slower one on macOS only, where sparse was a requirement.  Apparently with the new `sparseWriter` that’s no longer the case, we can have both.

I have no idea what’s true, and I don’t insist on switching to always-sparse here and removing the platform dependency, but I’d prefer the comment to document as much of the intent and trade-offs (even if it is “some previous version made that distinction for an unknown reason”) as possible, because our knowledge will not get better as time passes.

Reply:
> the implementation is now the same

I mean the decompression library being called.

Reply:
The AFAICT incorrect comment is still outstanding.

Reply:
What about:

```
        // Using special compressors on MacOS because generic
	// uses `io.Writer` that is less performant than `SparseWriter`
	// when writing applehv raw images.
```

Reply:
Let me try to be more clear:
- On macOS podman machine uses applehv
- applehv VMs are distributed as raw disk files (for windows hyperv we use the vhdx format)
- FCOS raw disk files have a size of 10GB with only ~2GB of actual files (vhdx has a size of ~2GB)
- gzip and zstd don't have any specific optimization for such kind of sparse files (the compressed size is ~8GB)
- As a result, the decompression of FCOS is particularly slow on macOS, 
- One way to make the decompression faster for these large sparse files is using SparseWriter
- The decompressor `generic` supports `gzip`, `zstd` etc...but doesn't use `SparseWriter`
- The zsdt and gzip decompressors use `SparseWriter` and are used only for FCOS raw disk files (i.e. macOS)

My previous comment `macOS gzipped VM images are sparse` was not accurate. It's not related to gzip but to the fact that `applehv` VMs are imported from raw disk files. And decompressing such files can be optimized using SparseWriter.

@baude is this correct?